### DISCUSSION TITLE
Dioxus app architecture insights and tooling

### DISCUSSION THREAD

**Discussion by gpoblon** [DioxusLabs/dioxus#2985] [General]
Hi Dioxus team,

Lately I've felt a lot more limited by architecture decisions in this very new ecosystem (Rust + Dioxus in webdev) than by writing Dioxus code.

While your vision suggests that one end goal would be to provide a complete framework (with deploy coming) which is great, and while you seem to have extensive knowledge about software architecture, resources on the matter (applicable to the rust ecosystem) are hard to come by.
Software architecture is complex and rust/dioxus introduces its own set of rules and features, making it even harder to make architectural decisions.
You have a vision, you provided a solution, but how would you use it?

Deploy will answer some of these and I'd love to read more about a roadmap.
It also rises a few questions: how much will it be locking, or homemade vs provide third party integration for example.

In the meantime, about dioxus apps architecture in general:

When building an application, Dioxus users will face a lot of questions and trade-offs choices regarding architecture.
Providing first-party crates/features/integration and insights —whether through articles, guides, advanced examples or even design suggestions— on aspects such as code structure, folder organization, and server-level / devops considerations would be incredibly valuable.
It would definitely help open the path toward scalable architectures for developers adopting Rust and Dioxus.

So what is this post about:
1. Having a place where the community can share knowledge, insights, guidance, and tooling related to architecture could be a good starting point.
2. But more importantly again, Dioxus team could bring incredibly valuable insights by sharing its vision and knowledge: posts or more official articles (I'd definitely buy your book!) presenting the blueprint of a Dioxus application that meets the requirements of modern architecture's characteristics, principles, structure, and rules, addressing among others the following topics:
  - [ ] Project structure
  - [ ] How you would handle monolithic/microservices (w/ services inter-communication) with scalability in mind
  - [ ] Observability
  - [ ] CI/CD
  - [ ] Testing and error-handling
  - [ ] Deployment
  - [ ] Dioxus, obviously

Its really about insights and guidance. It wouldn't have to force users into a specific direction or be turned into a tutorial.

For instance, I likely wouldn't approach Rust microservices the way I would in other contexts. I think Rust provides the ability to build monolithic apps that can be microservice-ready by splitting modules into crates within a workspace, or by leveraging Rust's features in a single crate, then compile it as a whole or separately (I'm curious to know which method you'd pick). How would you handle inter services communication (also server_fns)?

I hope this post will serve as a starting point for architecture discussions. It was so close to being named "Evan-driven architecture pls"!

**Comment by aynaash** [comment]
Yeah I am looking forward to this project's success and wide adoption. I think taking the route of ruby on rails and next for faster development while giving room for configuration would an amazing thing. I hope it all works out because I love Rust and would love to build everything with it I guess.
